service/src/main/java/bio/terra/cbas/controllers/RunSetsApiController.java

```diff
@@ -63,13 +73,47 @@ public RunSetsApiController(
 
   @Override
   public ResponseEntity<RunSetStateResponse> postRunSet(RunSetRequest request) {
-    if (!requestIsValid(request, this.cbasApiConfiguration)) {
-      return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
+    Gson gson = new Gson();
+
+    // request validation
+    Optional<ResponseEntity<RunSetStateResponse>> validateRequestResponse =
+        validateRequest(request, this.cbasApiConfiguration.getRunSetsMaximumRecordIds());
+    if (validateRequestResponse.isPresent()) {
+      return validateRequestResponse.get();
     }
 
-    // Store the method
-    UUID methodId = UUID.randomUUID();
+    // Fetch WDS Records and keep track of errors while retrieving records
+    String recordType = request.getWdsRecords().getRecordType();
+    List<String> recordIds = request.getWdsRecords().getRecordIds();
+    recordRecordsInRequest(recordIds.size());
 
+    ArrayList<RecordResponse> recordResponses = new ArrayList<>();
+    HashMap<String, String> recordIdsWithError = new HashMap<>();
+    for (String recordId : recordIds) {
+      try {
+        recordResponses.add(wdsService.getRecord(recordType, recordId));
```

Comment:
TOL: I suspect doing this as a stream would allow us to potentially do things in parallel rather than iterating one by one. A thought for "if we need to optimize", probably...

Reply:
I had initially tried to do this as a stream, but since `wdsService.getRecord` throws an exception it was becoming hard to figure it all out inside the map function because it didn't like checked exceptions being thrown inside its body. I am happy to mob on this though.